nvtabular/io/parquet.py

```diff
@@ -109,6 +132,458 @@ def to_ddf(self, columns=None):
+        file_sizes = defaultdict(int)
+        for rg in range(metadata.num_row_groups):
+            row_group = metadata.row_group(rg)
+            path = row_group.column(0).file_path
+            total_byte_size = row_group.total_byte_size
+            if total_byte_size > max_rg_size:
+                max_rg_size = total_byte_size
+                max_rg_size_path = path
+            file_sizes[path] += total_byte_size
+
+        # Check if any row groups are prohibitively large.
+        # Also check if any row groups are larger than recommended.
+        if max_rg_size > row_group_max_size:
+            # One or more row-groups are above the "required" limit
+            msg = (
+                f"Excessive row_group size ({max_rg_size}) detected in file "
+                f"{max_rg_size_path}. Please use the regenerate_dataset utility "
+                f"to rewrite your dataset."
+            )
+            warnings.warn(msg)
+        else:
+            # The only way size_valid==True is if we get here
+            size_valid = True
+
+        # Check if any files are smaller than the desired size.
+        # We only warn if there are >1 files in the dataset.
+        for path, size in file_sizes.items():
+            if size < file_min_size and len(pa_dataset.pieces) > 1:
+                msg = (
+                    f"File {path} is smaller than the desired dataset "
+                    f"partition size ({self.part_size}). Consider using the "
+                    f"regenerate_dataset utility to rewrite your dataset with a smaller "
+                    f"number of (larger) files."
+                )
+                warnings.warn(msg)
+                size_valid = False
+
+        # If the _metadata file is missing, we need to write
+        # it (or inform the user that it is missing)
+        if not metadata_file_exists:
+            if add_metadata_file:
+                # Write missing _metadata file
+                fs = self.fs
+                metadata_path = fs.sep.join([root_dir, "_metadata"])
+                with fs.open(metadata_path, "wb") as fil:
+                    metadata.write_metadata_file(fil)
+                meta_valid = True
+            else:
+                # Inform user that the _metadata file is missing
+                msg = (
+                    "For best performance with NVTabular, there should be a "
+                    "global _metadata file located in the root directory of the "
+                    "dataset. Please pass add_metadata_file=True to add the "
+                    "missing file."
+                )
+                warnings.warn(msg)
+                if require_metadata_file:
+                    meta_valid = False
+
+        # Return True if we have a parquet dataset with a _metadata file (meta_valid)
+        # and the row-groups and file are appropriate sizes (size_valid)
+        return meta_valid and size_valid
+
+    @classmethod
+    def regenerate_dataset(
```

Comment:
Whats the advantage of this regenerate_dataset method over something like ```nvtabular.Dataset(input_paths).to_parquet(output_paths)``` ?

Reply:
Yeah, good question - `Dataset.to_parquet` will produce `out_files_per_proc * N` output files (where `N` is the number of worker processes).  If you are using >1 worker, it is not possible to preserve the original dataset ordering.  The problem is that `to_parquet` relies on a `ParquetWriter` object being cached on the worker in betweeen IO tasks.  We avoid this problem in `regenerate_dataset`, because  we use CPU memory for the read.  After reading, we also aggregate the output files in CPU memory, allowing us to preserve the original ordering of the data (even when written in parallel).  This is because the "regenerate" version will always use CPU memory for the read.

With the above in mind, I agree that we could probably expand `to_parquet` to cover our needs.